### PR TITLE
Fix: Dandelion mutex LOCK usages

### PR DESCRIPTION
**Pull request by SmartArray**
This PR fixes the `stempool` mutex usage in `ActivateBestChain()`.

In some places the mutex wasn't acquired which could lead to race conditions, so it was added where needed.
And in some locations the mutex wasn't required where `EXCLUSIVE_LOCKS_REQUIRED` as modified accordingly.

### CODE REVIEW

**Comment by barrystyle** [comment]
Will fire this up now with `--enable-debug`

File: src/validation.cpp
Comment (barrystyle) [inline src/validation.cpp:2648]: not sure if this is needed, 'MaybeUpdateMempoolForReorg' contains an assert for stempool already

**Comment by barrystyle** [comment]
Was any of this backed with 'deadlocked thread' output from debug?
A lot of the locks i had added after the new implementation were suggested by the actions compiler in github.

**Comment by JaredTate** [comment]
ACK. Client compiles, runs & sends tx's fine. 

Side note: Still seeing the "Blockpolicy error mempool tx *** already being tracked" quite often.

**Review by JaredTate — APPROVED** [review]
ACK. Client compiles, runs & sends tx's fine.

Side note: Still seeing the "Blockpolicy error mempool tx *** already being tracked" quite often.

**Review by gto90 — APPROVED** [review]
cACK.  I would like to get @barrystyle 's questions answered before we merge.

File: src/validation.cpp
Comment (digicontributer) [reply src/validation.cpp:2648]: Yeah agree with @barrystyle here, seems redundant.

**Comment by digicontributer** [comment]
cACK for now.

File: src/validation.cpp
Comment (gto90) [reply src/validation.cpp:2648]: @SmartArray ,

Are you able to remove the redundant asserts?

**Comment by SmartArray** [comment]
> ACK. Client compiles, runs & sends tx's fine.
> 
> Side note: Still seeing the "Blockpolicy error mempool tx *** already being tracked" quite often.

This is seriously an issue we have to fix. @barrystyle Maybe you can take a look into this?
Many features are not working as expected, such as Mempool Resurrection after Reorg, and other things.

File: src/validation.cpp
Comment (SmartArray) [reply src/validation.cpp:2648]: I see where you are coming from. The signature of `MaybeUpdateMempoolForReorg`requires callers to assert that locks are held already. Super useful for static analysis.

This is also the reason the Bitcoin Developers added a check for the mempool mutex in the first place (L2647) although it is checked again in `MaybeUpdateMempoolForReorg`.

**Comment by SmartArray** [comment]
Would love to merge this if no one is opposed to that

**Review by digicontributer — APPROVED** [review]

**Pull request merged**
